### ISSUE TITLE
Questions About Reproducing the Released LM1B FMLM Checkpoint

### ISSUE THREAD

**Issue by bc-li** [david3684/flm#7]
Hello Chanhyuk,

Thank you for releasing the code and checkpoints. I am trying to reproduce the released `lm1b_fmlm.ckpt`, but the checkpoint metadata, the current released training script, and my reproduction results do not seem to match. I would really appreciate your help clarifying the exact training/evaluation recipe used for the released LM1B FMLM checkpoint.

From the current released code, LM1B FMLM distillation appears to use:

- Script: `scripts/train_lm1b_fmlm_denoiser.sh`
- Config: `algo=fmlm`, `model=small`, `data=lm1b-wrap`, `model.length=128`
- Distillation: `algo.distillation_method=PSD`
- PSD options: `diagonal_fraction=0.5`, `add_boundary=fixed`, `boundary_prob=32`, `offdiagonal_sampling=uniform_diff`
- Teacher: `algo.teacher_path=${TEACHER_PATH}`
- Student init in script/config: `algo.initialize_student_from_teacher=True`
- Optimizer in script: `optim.lr=3e-4`, `optim.beta2=0.95`
- Training length in script: `trainer.max_steps=1000000`

In the current implementation, FMLM PSD off-diagonal targets are built by forwarding the student at `(s,u)` and `(u,t)`, composing the two predictions, applying stop-gradient, and training the direct `(s,t)` prediction against that target.

However, the released `lm1b_fmlm.ckpt` metadata I read locally looks different:

```text
global_step = 550000
epoch = 39

algo.name = dos_denoiser
training.loss_type = sd
sampling.noise_removal = uniform_alpha
algo.use_discrete_schedule = true
algo.time_condition = alpha_t
algo.loss_scale = none

algo.distillation_method = PSD
algo.diagonal_fraction = 0.5
algo.offdiagonal_sampling = uniform_diff
algo.add_boundary = fixed
algo.boundary_prob = 32
algo.initialize_student_from_teacher = false
algo.use_teacher_for_D_t_lsd = false
algo.use_mse_loss_psd = false
algo.rescale_offdiag_loss_psd = false

training.ema = 0.9999, 0.999, 0.99, 0.95
optim.lr = 0.0003
optim.beta2 = 0.999
loader.global_batch_size = 512
loader.batch_size = 128
```

The metadata also records:

```text
algo.teacher_path = /home/david3684/ours_flow_ce_1m.ckpt
```

The checkpoint does not contain newer fields such as `use_ema_for_psd_target`, `psd_target_eval_mode`, or `set_midpoint`.

Using the same one-step evaluation protocol with `gamma=0.8`, GPT2-large GenPPL, and the BERT tokenizer:

```text
released lm1b_fmlm.ckpt:
  T1 GenPPL ~= 115.49
  entropy   ~= 4.145

my reproduction run:
  best local checkpoint around 325k:
    T1 GenPPL ~= 128.59
    entropy   ~= 4.166
  later checkpoints degraded again
```

This run(my reproduction run) used `algo=fmlm`, `training.loss_type=flow`, `sampling.noise_removal=ancestral`, `initialize_student_from_teacher=False`, `optim.beta2=0.999`, `global_batch_size=512`, and the released `lm1b_flm.ckpt` as the teacher.

I also tried a closer `sd + uniform_alpha + PSD` reproduction. Some checkpoints got lower GenPPL than the released checkpoint, especially at larger NFE, but the samples had much lower entropy, lower lexical diversity, more repetition, and more special-token / punctuation artifacts. This looked like distribution sharpening rather than matching the released checkpoint behavior.

So my questions are: 

1. Could you please confirm whether the released `lm1b_fmlm.ckpt` was trained with the current public `scripts/train_lm1b_fmlm_denoiser.sh`, or with an older/private `dos_denoiser` recipe?

2. Is `algo.name=dos_denoiser` exactly equivalent to the current `algo=fmlm` implementation? If not, could you explain the important differences?

3. Should LM1B FMLM distillation use `initialize_student_from_teacher=True` or `False`? The public script uses `True`, but the released checkpoint metadata says `False`.

4. Which optimizer/training length should reproduce the released checkpoint: `beta2=0.95, max_steps=1000000` from the public script, or `beta2=0.999, global_step=550000` from the checkpoint metadata?

5. For PSD off-diagonal targets, should `D_su` and `D_ut` be produced by the current student, an EMA student, or another teacher? Should this target branch run in train mode or eval mode?

6. Is `/home/david3684/ours_flow_ce_1m.ckpt` the same model as the released `lm1b_flm.ckpt`? If not, could you please share which FLM teacher checkpoint should be used?

7. What exact evaluation command/settings were used for the reported LM1B FMLM numbers, especially `gamma`, EMA decay, NFE, number of samples, and GenPPL scorer?

If possible, could you provide the original command/config or commit used to train and evaluate the released `lm1b_fmlm.ckpt`? Thanks again for your help!!

**Comment by david3684** [comment]
Hi,
Thanks for the detailed write-up. The field names you read from the metadata (`algo.name=dos_denoiser`, `training.loss_type=sd`, `sampling.noise_removal=uniform_alpha`, `algo.time_condition=alpha_t`, `algo.use_discrete_schedule`, etc.) come from our internal development repo, and those names are not in the public repo. They were all renamed and mapped into the current public `algo=fmlm` path, so the public `fmlm` implementation already includes that exact behavior. One thing to note is that `training.loss_type` and `sampling.noise_removal` aren't actually used anywhere for F(M)LM , so whatever you pass there has no effect on the result. 

Q1, Q2 : It was trained on private repo for the development, but it should act as same as `algo=fmlm`. 
Q3, Q4 : The checkpoint is trained with `False`, with `beta2=0.999`, while with `true` and `beta2=0.95` gives similar best results in our experience. Also, for the LM1B dataset I experienced quite a bit of fluctuation in gen.ppl / entropy over the training so recommend to add more eval iterations to fetch best ckpt. 
Q5: We didn't use EMA for stopgrad target. It should be generated as student running model. 
Q6: It is the same ckpt.
Q7: One-step (T1) generation is not affected by `gamma`. `gamma` only injects stochasticity for multi-step (>2 step) generation, so your T1 comparison is already fine on that front. For multi-step we reported the best over a gamma sweep, generally between 0.8 and 1.0. EMA decay was always 0.9999, and with 1024 samples + GPT-2 Large.

**Comment by david3684** [comment]
Since there's been no follow-up for a while, I'll close this for now.

**Comment by bc-li** [comment]
Thanks for the detailed clarification, and sorry for the late reply — I missed the notification. Really appreciate your help!
